backends/tfhe-cuda-backend/cuda/src/integer/scalar_mul.cuh

```diff
@@ -170,4 +172,134 @@ __host__ void host_integer_small_scalar_mul_radix(
     output_lwe_array->degrees[i] = input_lwe_array->degrees[i] * scalar;
   }
 }
+
+template <typename Torus>
+__host__ uint64_t scratch_cuda_integer_radix_scalar_mul_high_kb(
+    cudaStream_t const *streams, uint32_t const *gpu_indexes,
+    uint32_t gpu_count, int_scalar_mul_high<Torus> **mem_ptr,
+    uint32_t num_radix_blocks, int_radix_params params,
+    bool anticipated_buffer_drop, bool allocate_gpu_memory) {
+
+  uint64_t size_tracker = 0;
+
+  *mem_ptr = new int_scalar_mul_high<Torus>(
+      streams, gpu_indexes, gpu_count, params, num_radix_blocks,
+      allocate_gpu_memory, LEFT_SHIFT, anticipated_buffer_drop, &size_tracker);
+
+  return size_tracker;
+}
+
+template <typename Torus>
+__host__ void host_integer_radix_scalar_mul_high_kb(
+    cudaStream_t const *streams, uint32_t const *gpu_indexes,
+    uint32_t gpu_count, CudaRadixCiphertextFFI *ct,
+    int_scalar_mul_high<Torus> *mem_ptr, Torus *const *ksks, uint64_t rhs,
+    uint64_t const *decomposed_scalar, uint64_t const *has_at_least_one_set,
+    CudaModulusSwitchNoiseReductionKeyFFI const *ms_noise_reduction_key,
+    void *const *bsks, uint32_t num_scalars) {
+
+  if (rhs == (uint64_t)0) {
+    set_zero_radix_ciphertext_slice_async<Torus>(streams[0], gpu_indexes[0], ct,
+                                                 0, ct->num_radix_blocks);
+
+    return;
+  }
+
+  CudaRadixCiphertextFFI *tmp_ffi = mem_ptr->tmp;
+
+  copy_radix_ciphertext_slice_async<Torus>(streams[0], gpu_indexes[0], tmp_ffi,
+                                           0, ct->num_radix_blocks, ct, 0,
+                                           ct->num_radix_blocks);
+
+  bool allCarriesAreEmpty = true;
+  for (uint32_t i = 0; i < tmp_ffi->num_radix_blocks; ++i) {
+    if (tmp_ffi->degrees[i] >= mem_ptr->params.message_modulus) {
+      allCarriesAreEmpty = false;
+      break;
+    }
+  }
+
+  if (!allCarriesAreEmpty) {
+    host_full_propagate_inplace(streams, gpu_indexes, gpu_count, tmp_ffi,
+                                mem_ptr->fullprop_mem, (uint64_t **)ksks,
+                                ms_noise_reduction_key, bsks,
+                                tmp_ffi->num_radix_blocks);
+  }
```

Comment:
For all other functions this check is done on the Rust side, and actually it's ok if we keep it there (at least for now), to be compliant with the rest of the code. Could you then remove it from here and add it back on the Rust side please? :slightly_smiling_face: